File: backend/app/schemas/worksheet.py
from typing import Literal

from pydantic import BaseModel, Field


ContestLevel = Literal["AMC 10", "AMC 12", "AIME", "UKMT Senior", "IMO"]
GenerationStep = Literal["research", "questions", "solutions", "all"]
JobStatus = Literal["queued", "running", "completed", "failed"]


class WorksheetRequest(BaseModel):
    topic: str = Field(..., min_length=1, max_length=200)
    subtopics: list[str] = Field(..., min_length=1)
    step: GenerationStep = "all"
    core_questions_per_subtopic: int = Field(default=5, ge=1, le=20)
    challenge_questions_per_subtopic: int = Field(default=3, ge=0, le=20)
    tech_active_questions_per_subtopic: int = Field(default=1, ge=0, le=10)
    contest_level: ContestLevel = "IMO"
    diagram_mode: bool = False
    diagrams_per_subtopic: int = Field(default=1, ge=0, le=5)


class WorksheetOutput(BaseModel):
    research: str | None = None
    worksheet: str | None = None
    solutions: str | None = None


class WorksheetJob(BaseModel):
    id: str
    status: JobStatus
    topic: str
    step: GenerationStep
    error: str | None = None
    outputs: WorksheetOutput | None = None
    created_at: str
    updated_at: str

File: frontend/src/types/worksheet.ts
export type ContestLevel =
  | "AMC 10"
  | "AMC 12"
  | "AIME"
  | "UKMT Senior"
  | "IMO";

export type GenerationStep = "research" | "questions" | "solutions" | "all";

export type JobStatus = "queued" | "running" | "completed" | "failed";

export interface WorksheetRequest {
  topic: string;
  subtopics: string[];
  step: GenerationStep;
  core_questions_per_subtopic: number;
  challenge_questions_per_subtopic: number;
  tech_active_questions_per_subtopic: number;
  contest_level: ContestLevel;
  diagram_mode: boolean;
  diagrams_per_subtopic: number;
}

export interface WorksheetOutput {
  research: string | null;
  worksheet: string | null;
  solutions: string | null;
}

export interface WorksheetJob {
  id: string;
  status: JobStatus;
  topic: string;
  step: GenerationStep;
  error: string | null;
  outputs: WorksheetOutput | null;
  created_at: string;
  updated_at: string;
}

export interface WorksheetFormState {
  topic: string;
  subtopics: string[];
  step: GenerationStep;
  coreQuestionsPerSubtopic: number;
  challengeQuestionsPerSubtopic: number;
  techActiveQuestionsPerSubtopic: number;
  contestLevel: ContestLevel;
  diagramMode: boolean;
  diagramsPerSubtopic: number;
}

export const defaultFormState: WorksheetFormState = {
  topic: "",
  subtopics: [""],
  step: "all",
  coreQuestionsPerSubtopic: 5,
  challengeQuestionsPerSubtopic: 3,
  techActiveQuestionsPerSubtopic: 1,
  contestLevel: "IMO",
  diagramMode: false,
  diagramsPerSubtopic: 1,
};

export function toWorksheetRequest(form: WorksheetFormState): WorksheetRequest {
  return {
    topic: form.topic.trim(),
    subtopics: form.subtopics.map((s) => s.trim()).filter(Boolean),
    step: form.step,
    core_questions_per_subtopic: form.coreQuestionsPerSubtopic,
    challenge_questions_per_subtopic: form.challengeQuestionsPerSubtopic,
    tech_active_questions_per_subtopic: form.techActiveQuestionsPerSubtopic,
    contest_level: form.contestLevel,
    diagram_mode: form.diagramMode,
    diagrams_per_subtopic: form.diagramsPerSubtopic,
  };
}
